Add GoCD helm chart

GaneshSPatil opened the pull request.

GaneshSPatil committed:
Add GoCD helm chart

GaneshSPatil committed:
Fix lint

unguiculus commented:
Thanks for the PR. Please make sure you add standard labels. Also, we've started namespacing templates (see #1785). If you create a new chart with `helm create` and Helm 2.7+, you getll an initial experience with current best practices applied and a `_helpers.tpl` with namespaced templates.

GaneshSPatil committed:
Define standard labels and define 'gocd.name' and 'gocd.fullname' under '_helpers.tpl'
Add description about all the properties in ‘values.yaml’
specify a selector for gocd server and agent deployment

GaneshSPatil commented:
@unguiculus -- Thanks for the feedback.  I have made the requested changes to the PR.
If there anything specific that you would like to address, please let me know.

unguiculus commented on incubator/gocd/README.md:
```diff
@@ -0,0 +1,184 @@
+
+## Prerequisites
+
+- Kubernetes 1.5+ with Beta APIs enabled
```
1.7+ is officially supported.

unguiculus commented on incubator/gocd/templates/_helpers.tpl:
```diff
@@ -0,0 +1,20 @@
+Expand the name of the chart.
+*/}}
+{{- define "gocd.name" -}}
+{{- default .Chart.Name | trunc 24 | trimSuffix "-" -}}
```
You should truncate to 63 chars.

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "gocd.name" . }}-agent-deployment
```
`name: {{ template "gocd.fullname" . }}-agent`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+metadata:
+  name: {{ template "gocd.name" . }}-agent-deployment
+  labels:
+    app: {{ template "gocd.name" . }}-agent
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: agent
```
`component: agent`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+  replicas: {{ .Values.agent.replicaCount }}
+  selector:
+    matchLabels:
+      app: {{ template "gocd.name" . }}-agent
```
`app: {{ template "gocd.name" . }}`
`release: {{ .Release.Name | quote }}`
`component: agent`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+    metadata:
+      labels:
+        app: {{ template "gocd.name" . }}-agent
+        role: agent
```
`component: agent`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+  template:
+    metadata:
+      labels:
+        app: {{ template "gocd.name" . }}-agent
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+metadata:
+  name: {{ template "gocd.name" . }}-server-deployment
+  labels:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+  {{- end }}
+  selector:
+    matchLabels:
+      app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`
`release: {{ .Release.Name | quote }}`
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-server-godata-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+metadata:
+  name: {{ .Values.server.persistence.godata.name }}
+  labels:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/gocd-server-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+metadata:
+  name: {{ .Values.server.persistence.homego.name }}
+  labels:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+metadata:
+  name: {{ template "gocd.name" . }}-server-ingress
+  labels:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/service.yaml:
```diff
@@ -0,0 +1,31 @@
+metadata:
+  name: {{ template "gocd.name" . }}-server-service
+  labels:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`

unguiculus commented on incubator/gocd/templates/test-config.yaml:
```diff
@@ -0,0 +1,17 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name:  "{{ template "gocd.name" . }}-tests"
```
Add standard labels.

unguiculus replied:
`name:  "{{ template "gocd.fullname" . }}-tests"`

unguiculus commented on incubator/gocd/templates/tests/gocd-test.yaml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: "{{ template "gocd.name" . }}-test-{{ randAlphaNum 5 | lower }}"
```
`"{{ template "gocd.fullname" . }}-test-{{ randAlphaNum 5 | lower }}"`

unguiculus replied:
Add standard labels

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: {{ template "gocd.name" . }}-server-deployment
```
`name: {{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: server
```
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+    metadata:
+      labels:
+        app: {{ template "gocd.name" . }}-server
+        role: server
```
`app: {{ template "gocd.name" . }}`
`release: {{ .Release.Name | quote }}`
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-server-godata-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: server
```
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,96 @@
+  name: {{ template "gocd.name" . }}-agent-deployment
+  labels:
+    app: {{ template "gocd.name" . }}-agent
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/gocd-agent-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+  name: {{ .Values.agent.persistence.homego.name }}
+  labels:
+    app: {{ .Values.agent.name }}
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+  name: {{ template "gocd.name" . }}-server-deployment
+  labels:
+    app: {{ template "gocd.name" . }}-server
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/gocd-server-godata-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+  name: {{ .Values.server.persistence.godata.name }}
+  labels:
+    app: {{ template "gocd.name" . }}-server
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/gocd-server-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+  name: {{ .Values.server.persistence.homego.name }}
+  labels:
+    app: {{ template "gocd.name" . }}-server
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+  name: {{ template "gocd.name" . }}-server-ingress
+  labels:
+    app: {{ template "gocd.name" . }}-server
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/service.yaml:
```diff
@@ -0,0 +1,31 @@
+  name: {{ template "gocd.name" . }}-server-service
+  labels:
+    app: {{ template "gocd.name" . }}-server
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
```
`chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"`

unguiculus commented on incubator/gocd/templates/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: server
```
`component: server`

unguiculus commented on incubator/gocd/templates/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: {{ template "gocd.name" . }}-server-ingress
```
`name: {{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+    {{- end }}
+spec:
+  backend:
+    serviceName: {{ template "gocd.name" . }}-server-service
```
`{{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/service.yaml:
```diff
@@ -0,0 +1,31 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "gocd.name" . }}-server-service
```
`name: {{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/service.yaml:
```diff
@@ -0,0 +1,31 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: server
```
`component: server`

unguiculus commented on incubator/gocd/templates/service.yaml:
```diff
@@ -0,0 +1,31 @@
+      protocol: TCP
+      name: https
+  selector:
+    app: {{ template "gocd.name" . }}-server
```
`app: {{ template "gocd.name" . }}`
`release: {{ .Release.Name | quote }}`
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-server-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+    chart: "{{ template "gocd.name" . }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    role: server
```
`component: server`

unguiculus commented on incubator/gocd/templates/gocd-server-godata-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+  storageClassName: {{ .Values.server.persistence.godata.storageClass }}
+{{- else }}
+  storageClassName: ""
+  selector:
```
Selectors not needed.

GaneshSPatil replied:
@unguiculus -- GoCD server needs 2 Volumes, one to store `godata` and the other one for `homego`. Without selectors, is there any way of binding of a PVC to a specific persistent volume?

unguiculus replied:
There is not a single chart that uses selectors on PVCs. However, they may be necessary when you want to specify an existing PV. Otherwise, when the PV is auto-provisioned you should be fine without selectors. If you do want to allow selectors, then do it like so:

```yaml
{{- if .Values.server.persistence.godata.pvSelector }}
  selector: 
{{- toYaml .Values.server.persistence.godata.pvSelector | indent 4 }}
{{- end }}
```

Apply this for all PVCs.

unguiculus commented on incubator/gocd/templates/gocd-server-godata-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: {{ .Values.server.persistence.godata.name }}
```
`name: {{ template "gocd.fullname" . }}-data`

unguiculus commented on incubator/gocd/templates/gocd-server-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: {{ .Values.server.persistence.homego.name }}
```
`name: {{ template "gocd.fullname" . }}-home`

unguiculus commented on incubator/gocd/templates/gocd-agent-homego-pvc.yaml:
```diff
@@ -0,0 +1,26 @@
+  storageClassName: {{ .Values.agent.persistence.homego.storageClass }}
+{{- else }}
+  storageClassName: ""
+  selector:
```
Selectors not needed.

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+      volumes:
+        - name: godata-vol
+          persistentVolumeClaim:
+            claimName: {{ .Values.server.persistence.godata.name }}
```
`claimName: {{ template "gocd.fullname" . }}-data`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,72 @@
+            claimName: {{ .Values.server.persistence.godata.name }}
+        - name: homego-vol
+          persistentVolumeClaim:
+            claimName: {{ .Values.server.persistence.homego.name }}
```
`claimName: {{ template "gocd.fullname" . }}-home`

unguiculus reviewed (changes requested):
I made quite some comments. It's basically all about naming. I hope it's not too confusing.

unguiculus commented:
Please consider namespacing templates. See #1785. This should at least be done before a move to stable.

GaneshSPatil committed:
Fix comments
* Use namespaced templates
* Add standard labels
* Fix selectors on all PVCs
* Update README.md to reflect the changes

unguiculus commented on incubator/gocd/templates/_helpers.tpl:
```diff
@@ -0,0 +1,20 @@
+
+{{/*
+Create a default fully qualified app name.
+We truncate at 24 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
```
24 -> 63

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,99 @@
+metadata:
+  name: {{ template "gocd.fullname" . }}-agent
+  labels:
+    app: {{ template "gocd.fullname" . }}-agent
```
`app` label should be `app: {{ template "gocd.name" . }}` everywhere.

unguiculus commented on incubator/gocd/templates/_helpers.tpl:
```diff
@@ -0,0 +1,20 @@
+{{- default .Chart.Name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{- define "gocd.version" -}}
```
You don't seem to use this.

unguiculus commented on incubator/gocd/templates/gocd-agent-homego-pvc.yaml:
```diff
@@ -0,0 +1,27 @@
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    component: agent
```
Component should match name suffix.

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,75 @@
+              port: 8153
+            initialDelaySeconds: {{ .Values.server.healthCheck.initialDelaySeconds }}
+          {{- if .Values.server.persistence.enabled }}
+          volumeMounts:
```
Have you thought about using only one PVC with different subPaths? Just asking...

GaneshSPatil committed:
Fix comments
    * Change description to truncate from 24 chars to 63 chars
    * Change app label from gocd.fullname to gocd.name
    * Remove unnecessary selectors
    * Add gocd.name suffix to all the component label
    * Disable ingress and server persistence by default
    * Use subpath for go-sever /home/go and for /godata persistence volumes

unguiculus commented:
Marking stale.

GaneshSPatil commented:
@unguiculus -- apology for delayed response.

I have fixed the comments and have used one PVC with different subPaths for GoCD Server `/godata` and `/homego` volumes.

GaneshSPatil committed:
Bump up appVersion to the latest GoCD Version '17.12.0'

GaneshSPatil committed:
Add support for custom scripts

GaneshSPatil committed:
Support specifying extra environment variables which are passed to the server
* This is useful to specify GOCD_PLUGIN_INSTALL_<plugin_name> environment variable

GaneshSPatil committed:
Modify agent volume mount to support subpaths
* Support init scripts for agent

GaneshSPatil commented:
@unguiculus -- We've released a newer version of GoCD (`v17.12.0`). In which we have introduced another volume mount to support `/docker-entrypoint.d` scripts. I have made according changes to the helm chart to support this option using subpaths.

GaneshSPatil commented:
/approve cancel

unguiculus commented on incubator/gocd/templates/NOTES.txt:
```diff
@@ -0,0 +1,19 @@
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "gocd.name" . }}-server-service -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.server.service.externalPort }}
+{{- else if contains "ClusterIP" .Values.server.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "gocd.name" . }}-server," -o jsonpath="{.items[0].metadata.name}")
```
Add release label to the selector.

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          {{- end }}
+          imagePullPolicy: {{ .Values.agent.image.pullPolicy }}
+          resources:
+          {{ toYaml .Values.server.resources | indent 12 }}
```
This should not be indented.

kishorenc commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,73 @@
+            - name: GO_SERVER_SYSTEM_PROPERTIES
+              value: {{ .Values.server.env.goServerSystemProperties }}
+            {{- end }}
+{{ toYaml .Values.server.env.extraEnvVars | indent 12 }}
```
@GaneshSPatil I tried running this chart on our cluster with a `goServerSystemProperties` value, and I faced an issue with this line. A`null` is spit out causing the generated YAML to be incorrect. To fix, I modified this line to:

```
{{- if .Values.server.env.extraEnvVars }}
  {{ toYaml .Values.server.env.extraEnvVars | indent 12 }}
{{- end }}
```

A snippet from the configuration I used:

```
server:
  # other configurations
  # ...
  env:
    goServerSystemProperties: "-Xmx1024m"
    extraEnvVars:
```

GaneshSPatil committed:
Fixed comments
* Add release label to the selector
* Indent resources values section
* Validate extraEnvVars value for not null

GaneshSPatil commented:
Thanks, @unguiculus @kishorenc. I have made changes to the helm chart accourdingly!

unguiculus commented on incubator/gocd/Chart.yaml:
```diff
@@ -0,0 +1,20 @@
+sources:
+- https://github.com/gocd/gocd
+maintainers:
+- name: ThoughtWorks, Inc.
```
Use Github username.

GaneshSPatil replied:
@unguiculus -- GoCD Helm Chart is done by [`GoCD`](https://github.com/gocd/gocd) development team. Everyone from the team will be maintaining this Helm chart.

Is there any specific reason for having Github username as the maintainer name?

mattfarina replied:
@GaneshSPatil There are a few reasons for using GitHub usernames.

1. It lets us identify who the chart owners are as people. We can see who comments and has an opinion on GitHub and if they are a chart maintainer or not. This is also useful when we cc people in issues to know the GitHub names to use.
2. We recently instituted OWNERS files on this repo. This has the ability for people to be notified for review of a PR automatically and to have the ability to merge PRs for this chart.  [I recently wrote about it](https://codeengineered.com/blog/2017/kubernetes-chart-owners/). Eventually we'd like to have folks start using an OWNERS file for this chart. The names in them are the GitHub usernames and to have that tracibility would be useful.

I fully respect that this is a company project by ThoughtWorks. It's just that we need a way to interact with people who are involved with maintaining the chart.

GaneshSPatil replied:
@mattfarina -- Thanks for the explanation, I will change the maintainers' names accordingly.

unguiculus commented:
/ok-to-test

GaneshSPatil committed:
Disable agent persistence by default

unguiculus commented:
/retest

unguiculus commented:
/retest

GaneshSPatil commented:
@unguiculus -- looks like it took more than 5 minutes to execute the GoCD helm chart tests. We'll take a look at the failed tests and retrigger the build.

GaneshSPatil committed:
Fix tests
* Fix configMap name under volume mounts

GaneshSPatil commented:
@unguiculus -- I've fixed the failing tests. Please take a look.

GaneshSPatil committed:
Change maintainers list to contain Github usernames of individuals instead company name

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+metadata:
+  name: {{ template "gocd.fullname" . }}-agent
+  labels:
+    app: {{ template "gocd.name" . }}-agent
```
The `app` label should be `{{ template "gocd.name" . }}` on all resources.

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+  template:
+    metadata:
+      labels:
+        app: {{ template "gocd.name" . }}-agent
```
Fix app label.

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,75 @@
+          {{- if .Values.server.image.tag }}
+          image: "{{ .Values.server.image.repository }}:{{ .Values.server.image.tag }}"
+          {{- else }}
+          image: "{{ .Values.server.image.repository }}:v{{ default .Chart.AppVersion .Values.app.version }}"
```
I don't think we should have an additional `app.version`. It's redundant. You should just use the image tag instead.

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+    chart:  "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
+    component: {{ template "gocd.name" . }}-agent
```
Don't include the name template in components. That's in the `app` label already.

`component: agent`

Please fix it throughout the chart.

GaneshSPatil committed:
Fix comments
* Fix app label to be 'app: {{ template gocd.name . }}' on all the resources
* Change component from 'component: {{ template gocd.name . }}-server' to 'component: server' on all the server resources
* Change component from 'component: {{ template gocd.name . }}-agent' to 'component: agent' on all the agent resources
* Remove app.version attribute. Rely on .Charts.appVersion or .Values.server.image.tag

unguiculus commented on incubator/gocd/templates/NOTES.txt:
```diff
@@ -0,0 +1,19 @@
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "gocd.name" . }}-server-service -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.server.service.externalPort }}
+{{- else if contains "ClusterIP" .Values.server.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "gocd.name" . }}-server,release={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
```
`-l "app={{ template "gocd.name" . }},release={{ .Release.Name }}"`

unguiculus commented on incubator/gocd/templates/NOTES.txt:
```diff
@@ -0,0 +1,19 @@
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.server.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "gocd.name" . }}-server-service'
```
`{{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/NOTES.txt:
```diff
@@ -0,0 +1,19 @@
+{{- else if contains "LoadBalancer" .Values.server.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "gocd.name" . }}-server-service'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "gocd.name" . }}-server-service -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```
`{{ template "gocd.fullname" . }}-server`

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,75 @@
+    heritage: {{ .Release.Service | quote }}
+    component: server
+spec:
+  {{- if le .Values.server.replicaCount 1.0 }}
```
You compare against a float value. This may fail if `replicaCount` is specified as string. Use the `int` function
`{{- if le (int .Values.server.replicaCount) 1 }}`.

But why are you doing this at all? This doesn't make sense to me.

GaneshSPatil replied:
@unguiculus -- we need to restrict the replicaCount of GoCD server to 1. Running more than one GoCD server will end up in an unexpected behavior of the application.

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+          {{- end }}
+          {{- if .Values.agent.persistence.enabled }}
+          volumeMounts:
+          - name: goagent-vol
```
Indent one more level and make sure list indentation is consistent.

unguiculus reviewed (changes requested):
A few more things. Sorry for the delay. I just realized that my review was still pending and I had forgotten to submit it.

GaneshSPatil committed:
Fix minor comments
* Rename gocd-server-service to gocd-server
* Reindent list indentation
* Cast repicaCount value to integer for comparison

GaneshSPatil commented:
@unguiculus -- is there anything else that needs to be done on the chart?

unguiculus commented on incubator/gocd/Chart.yaml:
```diff
@@ -0,0 +1,24 @@
+maintainers:
+- name: arvindsv
+  email: arvind@thoughtworks.com
+- name: GaneshSPatil
```
Use Github username

unguiculus commented on incubator/gocd/README.md:
```diff
@@ -0,0 +1,165 @@
+
+## Prerequisites
+
+- Kubernetes 1.7+ with Beta APIs enabled
```
1.8+

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,98 @@
+apiVersion: extensions/v1beta1
```
`apps/v1beta2`

unguiculus commented on incubator/gocd/templates/tests/gocd-test.yaml:
```diff
@@ -0,0 +1,42 @@
+        set -ex
+        # copy bats to tools dir
+        cp -R /usr/local/libexec/ /tools/bats/
+      volumeMounts:
```
List indentation inconsistent. Always indent two spaces to be consistent.

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,75 @@
+    component: server
+spec:
+  {{- if le (int .Values.server.replicaCount) 1 }}
+  replicas: {{ .Values.server.replicaCount }}
```
Why do you make this configurable at all? This doesn't seem to make sense.

GaneshSPatil replied:
Supported values for GoCD Server replicas Count are `0` and `1`. Where,
`0` indicates `Agent Only Deployment` and
`1` indicates `Agent-Server Deployment`.

One can choose to deploy only GoCD agents on the kubernetes cluster. In such cases the `server.replicaCount` value will be `0`.

unguiculus replied:
Is `0` a valid value? Didn't know that but seems strange to me. Why don't you add an if statement around the deployment? That would be the usual way how this is done. And make it a boolean value.

GaneshSPatil replied:
I've changed the `server.replicaCount` variable to `server.enabled` variable which can have boolean value
`true`: `agent-server deployment`  and
`false`: `agent-only deployment`.

GaneshSPatil committed:
Fix comments
* Use lowercase github username
* Bump up minimum supported kubernetes api version to 1.8+
* Change extensions/v1beta1 to apps/v1beta2
* Indent list

unguiculus commented on incubator/gocd/templates/gocd-agent-deployment.yaml:
```diff
@@ -0,0 +1,102 @@
+  replicas: {{ .Values.agent.replicaCount }}
+  selector:
+    matchLabels:
+      app: {{ template "gocd.name" . }}
```
Add release label to selector.

unguiculus commented on incubator/gocd/templates/gocd-server-deployment.yaml:
```diff
@@ -0,0 +1,79 @@
+  {{- end }}
+  selector:
+    matchLabels:
+      app: {{ template "gocd.name" . }}
```
Add release label to selector.

GaneshSPatil committed:
Fix comments
* Add release label to deployment selectors
* Change server.replicaCount to server.enabled for the agent-server and agent-only deployment mode

GaneshSPatil commented:
@unguiculus -- I've fixed all the existing comments on the PR. is there anything else that needs to be addressed?

GaneshSPatil commented:
Also, I see the `tide` check is pending on the PR. is there anything specific that needs to be done for it?

unguiculus commented:
List indentation is still not consistent. If you like you can directly move the chart to stable before we merge it.

mattfarina commented:
@GaneshSPatil No. That's part of the merge automation.

GaneshSPatil committed:
Fix list indentation, bump up app version to 18.1.0.

GaneshSPatil commented:
> if you like you can directly move the chart to stable before we merge it.

@unguiculus -- We'd like to keep the helm chart under `incubator` for now. We'll send another pull request to move the chart to `stable`. 

Thank you for reviewing the PR.

unguiculus commented:
/lgtm
